### ISSUE TITLE
Creating a view inside an `effect` will track any signal reads that happen in the view

### ISSUE THREAD

**Issue by kfrancois** [angular/angular#54548]
### Which @angular/* package(s) are the source of the bug?

Don't known / other

### Is this a regression?

No

### Description

I have a `PopoverDirective` which contains an input that allows the developer to pass in a custom template. It works like this:
```ts
<div
  appPopover
  [isPopoverVisible]="isPopoverVisible()"
  [popoverTemplate]="popoverTemplate"
>
   Regular content here
</div>

<ng-template #popoverTemplate> Popover content here </ng-template>
```

Internally, this directive fill open or destroy the popover depending on the `isPopoverVisible` input:
```ts
effect(() => {
  if (this.isPopoverVisible()) {
    this.openPopover();
  } else {
    this.destroyPopover();
  }
});
```

This works welll, but when adding signals into the `popoverTemplate` I notice that any effects contained inside the `appPopover` directive will fire again. Even if none of the directive's inputs have changed.

For example:

```ts
<div
  appPopover
  [isPopoverVisible]="isPopoverVisible()"
  [popoverTemplate]="popoverTemplate"
>
   Regular content here
</div>

<ng-template #popoverTemplate> {{ someSignal() }} </ng-template>
```

When `someSignal` changes, my effect inside the `appPopover` directive will now fire _again_, even though none of its inputs have changed (only the input of the template we pass in has changed).

This feels unintentional, I'd like to know if this is intended behaviour, or a bug in the framework.

### Please provide a link to a minimal reproduction of the bug

https://github.com/kfrancois/angular-effects-repro

Also available in Stackblitz: https://stackblitz.com/github/kfrancois/angular-effects-repro

### Please provide the exception or error you saw

Using the above link, follow these steps:

1. Open the console
2. Click on the `Show popover` button
3. Click on the `Do async logic to populate popover template` button
4. Observe the console output

On the "working" example, `isPopoverVisible: true` does not log twice.
On the "broken" example, `isPopoverVisible: true` does log twice, which is unexpected.

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 17.2.0
Node: 21.6.1 (Unsupported)
Package Manager: npm 10.4.0
OS: darwin arm64

Angular: 17.2.1
```

**Comment by atscott** [comment]
This is working as expected. `openPopover` executes the popover template inside the `effect` which adds any signals in the template to the effect. You need to wrap `openPopover` inside an `untracked` call to prevent this: https://stackblitz.com/edit/github-kphkvx?file=src%2Fmain.ts,src%2Fapp%2Fbroken-popover.component.ts,src%2Fapp%2Fpopover-directive.ts,src%2Fapp%2Fworking-popover.component.ts

**Comment by alxhub** [comment]
Note that this isn't exactly as expected - one of the reasons `effect` is still developer preview is that its interaction with different framework APIs is still a work in progress. In particular, it's an open question whether Angular should untrack all rendering operations by default, or throw errors if you accidentally call them from reactive contexts, or just run the operations like it does today.

**Comment by kfrancois** [comment]
Understood! Thank you Andrew & Alex for providing a workaround, and for providing feedback on this issue. 😄 🙏 

Glad to work with the current behaviour for the time being, since as you mentioned `effect` is in dev preview. 


My personal feeling is that the current behaviour is unexpected as it adds a hard to debug pitfall that's not immediately noticeable (IE: devs expect `effects` callbacks only to run when a tracked signal changes).

The other options you mentioned both sound like valid paths to me (I'm curious about other developers' experience/opinions around either). Excited to follow along. 😄 


Feel free to repurpose/close this issue as desired!

**Comment by kfrancois** [comment]
tl;dr for anyone following along:

When creating a view (e.g. through CDK Overlay) inside an `effect` function, this `effect` function will track any signals inside the created view. The current solution is to create this view inside an `untracked(() => { ... })` function.

```ts

// Template passed into a view that will be instantiated
<ng-template #templateRef>
   {{ someSignal() }}
</ng-template>


// Effect that either instantiates or destroys the view
createViewEffect = effect(() => {
  if (this.isPopoverVisible()) {

    // When wrapping with `untracked`, we ensure that this effect does not track signals inside the view
    untracked(() => this.openPopover());
  } else {
    this.destroyPopover();
  }
});

openPopover() {
  const templateRef = this.templateRef();

  // Create an embedded view using this template
  ...
}
```
